Add support for environment variable injection in kibana.yml

fbaligand opened the pull request:
This PR lets support for environment variable injection in kibana.yml, using syntax `${MY_ENV_VAR}`.

Fixes #6688

elasticmachine commented:
Since this is a community submitted pull request, a Jenkins build has not been kicked off automatically. Can an Elastic organization member please verify the contents of this patch and then kick off a build manually?

tylersmalley commented on src/server/config/config.js:
```diff
@@ -61,13 +61,25 @@ export class Config {
     this._commit(config);
   }
 
+  _replaceEnvVars(value) {
```
Can we add tests for all of these cases in addition to the env variable not being specified?

fbaligand replied:
Hi @tylersmalley 

- OK, I will add a test when env variable is not specified.
- Then, for additional cases, currently, `_.isString(value)` and `_.isArray(value)` are covered. Only `_.isPlainObject(value)` is not covered. I will change the test to cover it.

fbaligand commented on src/server/config/config.js:
```diff
@@ -61,13 +61,25 @@ export class Config {
     if (_.isPlainObject(key)) {
       config = override(config, key);
```
@tylersmalley 
- Question : should I replace `config = override(config, key);` by `config = override(config, _replaceEnvVars(key));` ?

jbudz replied:
I'm not entirely sure what it's used for, maybe testing.  @spalger do you know?

spalger replied:
Not sure it is used. Gut feeling is it was just an "expected" behavior of a method like this.

fbaligand replied:
So, should I replace override(config, key) by override(config, _replaceEnvVars(key)) ?

spalger replied:
My remcommendation from https://github.com/elastic/kibana/pull/16988#issuecomment-379050543 is to move all of this into https://github.com/elastic/kibana/blob/master/src/cli/serve/read_yaml_config.js and do it once when reading `kibana.yml` from the file system

fbaligand replied:
@spalger 
OK, I will do so.
Thanks for this first review.

fbaligand commented:
@tylersmalley 
What behaviour do you wish when environment variable is not defined ?
Currently, it replaces env var reference by `undefined`

fbaligand commented:
Hi @tylersmalley,

- I increased test coverage (all 'replaceEnvVars()' lines are covered)
- I raise an error if referenced environment variable does not exist

Is it OK for you ?

fbaligand commented:
Hi @tylersmalley,
Could you review my changes ?

tylersmalley commented:
@fbaligand, thanks for the updates. Hope to get to reviewing this soon.

fbaligand commented:
I hope so ! :)

spalger commented:
Just took a cursory look, but think I'd prefer that this was implemented in https://github.com/elastic/kibana/blob/master/src/cli/serve/read_yaml_config.js so that replacement like this only occured when reading the config file. I would be quite surprised by this behavior if I was calling `config.set()` directly from just about anywhere.

fbaligand commented:
I just rebased my code to merge conflicts with master branch

fbaligand commented:
@spalger 
I just moved the code to `read_yaml_config.js` as requested.
Could you review the code ?

fbaligand commented:
Great @spalger !

=> @tylersmalley @jbudz could you review/approve the code ?

spalger commented:
Jenkins, test this

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1537/)

jbudz commented on src/cli/serve/read_yaml_config.js:
```diff
@@ -1,7 +1,16 @@
+function _replaceEnvVarRefs(val) {
+  return val.replace(/\$\{(\w+)\}/g, (match, envVarName) => {
+    if (process.env[envVarName] !== undefined) {
+      return process.env[envVarName];
```
I would prefer if we _.escaped this by default, and any configuration that needs it unescaped can do so.  Thoughts?

fbaligand replied:
@jbudz 
So you want to replace `return process.env[envVarName];` 
by `return _.escape(process.env[envVarName]);` ?

No problem for me, but can you explain me why ? I mean : at this point, the yaml file is already parsed.

Maybe it could be an issue for `server.basePath` option if user wants to use `&` character ?

spalger replied:
Yeah, I don't think we want to escape HTML entities in the config values. People using this feature need to take care that they inject the correct value and Joi will take care of ensuring that it's a valid value.

I was initially concerned about non-string values, but realized Joi also parses JSON values in an attempt to coerce the value to the desired type, so to set a boolean value you can do:

```yaml
some.boolean: "${MY_ENV_VAR}"
```

and
 
```sh
MY_ENV_VAR=true bin/kibana
```

And Joi will correctly convert the valid JSON so that

```js
config.get('some.boolean') === true
```

jbudz replied:
in practice I can't think of anything realistic, it's mostly how I operate.  I'l  defer on this.

fbaligand replied:
I agree with @spalger 
In an option like `elasticsearch.customHeaders`, people want that the final value is exactly what they type. And value can contain some special characters like `&`.

spalger commented on src/cli/serve/read_yaml_config.js:
```diff
@@ -19,6 +28,10 @@ export function merge(sources) {
       }
 
+      if (isString(val)) {
+    	  val = _replaceEnvVarRefs(val);
```
Looks like this line is mixing tabs and spaces. Also, would you mind just inlining the `_replaceEnvVarRefs()` function here or removing the leading `_` from it's name?

fbaligand commented:
@spalger 
You're right, there was a remaining tab, sorry for that, I thought that linting protects against this kind of mistake.
I removed the tab and removed the leading `_` to `replaceEnvVarRefs` method.
Is it OK for you ?

fbaligand commented:
Hi @spalger @jbudz 
I saw that the PR was in conflict with master => I rebased my code and migrated my test to the new test convention.

=> Could you launch jenkins tests to check everything is still OK ?

jbudz commented:
test it please

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1647/)

fbaligand commented:
Jenkins build is OK => nice !

=> @tylersmalley, could you approve the PR ?

tylersmalley commented:
6.x/6.3: 6d70217571c55ee2bdb565e45debe2fc58b6a2d2

tylersmalley commented:
Thanks @fbaligand!
